cmd/kubeadm/app/cmd/phases/workflow/runner_test.go

```diff
@@ -0,0 +1,278 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package workflow
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"testing"
+)
+
+func phaseBuilder(name string, phases ...Phase) Phase {
+	return SimplePhase{
+		Name:   name,
+		Short:  fmt.Sprintf("long description for %s ...", name),
+		Phases: phases,
+	}
+}
+
+func TestComputePhaseRunFlags(t *testing.T) {
+	var w = Runner{
+		Phases: []Phase{
+			phaseBuilder("foo",
+				phaseBuilder("bar"),
+				phaseBuilder("baz"),
+			),
+			phaseBuilder("qux"),
+		},
+	}
+
+	var usecases = []struct {
+		name          string
+		options       RunnerOptions
+		expected      map[string]bool
+		expectedError bool
+	}{
+		{
+			name:     "no options > all phases",
+			options:  RunnerOptions{},
+			expected: map[string]bool{"foo": true, "foo/bar": true, "foo/baz": true, "qux": true},
+		},
+		{
+			name:     "options can filter phases",
+			options:  RunnerOptions{PhaseFilter: []string{"foo/baz", "qux"}},
+			expected: map[string]bool{"foo": false, "foo/bar": false, "foo/baz": true, "qux": true},
+		},
+		{
+			name:     "options can filter phases - hierarchy is considered",
+			options:  RunnerOptions{PhaseFilter: []string{"foo"}},
+			expected: map[string]bool{"foo": true, "foo/bar": true, "foo/baz": true, "qux": false},
+		},
+		{
+			name:     "options can exclude phases",
+			options:  RunnerOptions{ExcludePhases: []string{"foo/bar", "qux"}},
+			expected: map[string]bool{"foo": true, "foo/bar": false, "foo/baz": true, "qux": false},
+		},
+		{
+			name:     "options can exclude phases - hierarchy is considered",
+			options:  RunnerOptions{ExcludePhases: []string{"foo"}},
+			expected: map[string]bool{"foo": false, "foo/bar": false, "foo/baz": false, "qux": true},
+		},
+		{
+			name: "exclude options have higher precedence than filter options",
+			options: RunnerOptions{
+				PhaseFilter:   []string{"foo"},     //  "foo", "foo/bar", "foo/baz" true
+				ExcludePhases: []string{"foo/bar"}, //  "foo/bar" false
+			},
+			expected: map[string]bool{"foo": true, "foo/bar": false, "foo/baz": true, "qux": false},
+		},
+		{
+			name:          "invalid filter option",
+			options:       RunnerOptions{PhaseFilter: []string{"invalid"}},
+			expectedError: true,
+		},
+		{
+			name:          "invalid exclude option",
+			options:       RunnerOptions{ExcludePhases: []string{"invalid"}},
+			expectedError: true,
+		},
+	}
+	for _, u := range usecases {
+		t.Run(u.name, func(t *testing.T) {
+			w.prepareForExecution()
```

Comment:
Should `w` be defined here? Just so it's not modified by a test case.

Reply:
done